### PR TITLE
add boilerplate auth server to create JWT tokens

### LINKED ISSUE #451: Create a guide for using hasura with passportjs
Should cover:
- Deploying Hasura
- Deploying passportjs (including postgres migrations for setting up the tables)
- Option 1: webhook
 - Sample code
 - Architecture diagram of how the query authentication works
- Option 2: JWT
 - Sample code
 - Architecture diagram of how the query authentication works

### PR DESCRIPTION
**Pull request by peterjgrainger**
Add a boilerplate for creating a jwt authentication server to provide tokens to be used with the graphQL engine.  See the docker-compose configuration and the README to see how to test.

close #451 

### Description
Creates the auth server described here: https://docs.hasura.io/1.0/graphql/manual/auth/jwt.html to create tokens to be decoded by the GraphQL server.

What component does this PR affect? 

- [ ] Server
- [ ] Console
- [ ] CLI
- [ ] Docs
- [x] Community Content
- [ ] Build System

Requires changes from other components? If yes, please mark the components:

- [ ] Server
- [ ] Console
- [ ] CLI
- [ ] Docs
- [ ] Community Content
- [ ] Build System

### Related Issue
n/a

### Solution and Design
Based very closely on [boilerplate with passportjs](https://github.com/hasura/graphql-engine/tree/master/community/boilerplates/auth-webhooks/passport-js) but without the webhook.  Uses an environment variable to set the JWT signing key.

Allows a user to signup (saves user in postgres) then login using that user.

JWT contains userID from the database their username and the required hasura claims including the role `user` and signs with JWT signing key

```javascript
{
        sub: '' + user.id,
        name: user.username,
        iat: Date.now() / 1000,
        "https://hasura.io/jwt/claims": {
          "x-hasura-allowed-roles": ["editor","user", "mod"],
          "x-hasura-user-id": '' + user.id,
          "x-hasura-default-role": "user",
          "x-hasura-role": "user"
        }
      }
```

### Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update
- [ ] Community content

### Checklist:
- [x] I have read the **[contributing guide](https://github.com/hasura/graphql-engine/blob/master/CONTRIBUTING.md)** and my code conforms to the guidelines.
- [ ] This change requires a change in the documentation. 
- [ ] I have updated the documentation accordingly.
- [ ] I have added required tests.

### CODE REVIEW

**Comment by revskill10** [comment]
Please don't use `knex`. Use a graphql client instead.

**Comment by peterjgrainger** [comment]
@revskill10 I was trying to be consistent with the other boilerplates such as https://github.com/hasura/graphql-engine/tree/master/community/boilerplates/auth-webhooks/passport-js 

Do you mean to set up the users table using the GraphQL client?  I was trying to make the process as easy to use as possible, it's only saving the user against a password?  Maybe I could use redux or other simple data store instead?

**Comment by peterjgrainger** [comment]
Maybe I could add migrations and use the hasura cli? https://docs.hasura.io/1.0/graphql/manual/migrations/index.html

**Comment by revskill10** [comment]
I think you can just deploy a staging server for hasura, the code only uses graphql instead. You can put static schema file in repo optionally.

**Comment by peterjgrainger** [comment]
@revskill10 updated as suggested.  Let me know if there is anything else that should be changed

File: community/boilerplates/jwt-authentication-server/controllers/user.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  if (errors) {
+    return res.status(400).json({'errors': errors});
+  }
+
+  try {
+    await User.query()
+                           .allowInsert('[username, password]')
+                           .insert({
+                             username: req.body.username,
+                             password: req.body.password
```
Comment (coco98) [inline community/boilerplates/jwt-authentication-server/controllers/user.js]: Shouldn't this password be salted/hashed and saved. Maybe use a standard password creating library? This library will be used during signup for inserting, and comparing passwords when logging in?

**Comment by coco98** [comment]
@peterjgrainger Thank you so much for your work! This will be very helpful to folks looking at getting started with a simple JWT auth server. 🎉

I'll be reviewing this along with @ecthiender and @arvi3411301.

Also, please do sign the CLA so that we can merge your PR into the repo!

**Comment by peterjgrainger** [comment]
@coco98 I signed the CLA multiple times but the check is still failing.

Also I'm not sure why the circleCI build failed?  It passed when I did the initial change

**Comment by emahuni** [comment]
@peterjgrainger, can you please try again, wanted to see this in the community.

**Comment by ctwhome** [comment]
is this finally finished? thanks a lot

**Comment by Stefmore02** [comment]
Closing this PR in response to [[4442](https://github.com/hasura/graphql-engine/pull/4442)].

**Pull request closed without merging**